### PR TITLE
Implement page closing

### PR DESCRIPTION
**Pull request by shulcsm**
Would be nice to have a way to just close the target but i couldn't figure it out.

### CODE REVIEW

**Comment by mattsse** [comment]
Thanks for this!
I wasn't sure how to handle a page close command, in particularly how to make sure that `PageInner` gets dropped.
However the a `Page.Close` event eventually destroys the target which drops the receiver half of the channel.
The problem right now is that the `Page.close` command returns before the target is destroyed, making it possible to send commands to the browser for a target that is being destroyed. 
This behavior is fine for now, but needs more work, either delaying the `Page.close` command until `Target.targetdestroyed` was fired or add a designated Handlermessage variant to remove the target as soon as the `Page.close` is fired.

File: src/page.rs [hunk 1]
```diff
@@ -261,8 +261,10 @@ impl Page {
         Ok(resp.result.node)
     }
 
-    pub async fn close(self) {
-        todo!()
+    /// Closes page
+    pub async fn close(self) -> Result<()> {
+        self.execute(CloseParams {}).await?;
```
Comment (mattsse) [inline src/page.rs:266]: This is the correct command, but it should be changed to `CloseParams::default()` to follow the convention used for empty commands in this project.

Comment (mattsse) [inline src/page.rs:266]: Can you add a short note that indicates that this `Tries to close page, running its `beforeunload` hooks, if any` as well as a reference to `CloseParams`

**Review by mattsse — CHANGES_REQUESTED** [review]

**Comment by shulcsm** [comment]
> Thanks for this!
> I wasn't sure how to handle a page close command, in particularly how to make sure that `PageInner` gets dropped.
> However the a `Page.Close` event eventually destroys the target which drops the receiver half of the channel.
> The problem right now is that the `Page.close` command returns before the target is destroyed, making it possible to send commands to the browser for a target that is being destroyed.
> This behavior is fine for now, but needs more work, either delaying the `Page.close` command until `Target.targetdestroyed` was fired or add a designated Handlermessage variant to remove the target as soon as the `Page.close` is fired.

I'm not sure if  waiting for `Target.targetdestroyed` is necessary if assuming target eventually gets cleaned up. `Page` gets consumed and if you are dealing with raw ids you are on your own.

I think it's more important to figure out a way to just destroy the page target since for majority of cases you don't care about unload and just want to clean up.

**Comment by mattsse** [comment]
> I'm not sure if waiting for Target.targetdestroyed is necessary if assuming target eventually gets cleaned up. Page gets consumed and if you are dealing with raw ids you are on your own.

Agree. But there is a potential race condition for following scenario:

```rust
let el =  page.find_element("input#searchInput").await?;
...
page.close().await?;
...
// this will fail because page is already closed.
el.click().await?;
```

This `el.click()` will fail regardless, but at the moment the reason it fails is because the click command is sent to the browser and the browser returns an error. This happens because the `Target` of the `Page` is not removed yet and is still able to receive events sent from an `Element`. Preferably the `el.click()` should fail with a `SendError` because the receiver should already be dropped at this point, preventing the overhead of sending a request that is guaranteed to fail.

> I think it's more important to figure out a way to just destroy the page target since for majority of cases you don't care about unload and just want to clean up.

This can be done with `Target.closeTarget` instead, that's how [puppeteer](https://github.com/puppeteer/puppeteer/blob/main/src/common/Page.ts#L1812) does it.

**Pull request merged**